### PR TITLE
V2: Introduce TALL Stack

### PR DESCRIPTION
**Pull request by luchaos**
Introduces [TALL stack](https://tallstack.dev/) to the existing application without compromising any existing features.
Coming full circle with https://github.com/RetroAchievements/RAWeb/issues/42 - how the years fly by.

**Quickstart**

See updated [README.md](https://github.com/RetroAchievements/RAWeb/tree/v2).

> **Note**
> Update your `.env` file according to `.env.example`

```shell
composer install
composer setup

# link storage ...
php artisan ra:storage:link
# ... or when using Sail (Docker Compose):
sail artisan ra:storage:link

npm install

# build frontend ...
npm run build 
# ... or hot reload during frontend development:
npm run dev
```

**Docker**

Compose setup has been replaced with [Laravel Sail](https://laravel.com/docs/sail).
Read `sail` commands as `./vendor/bin/sail` (you might want to add an alias to your favourite shell for that) [as mentioned in Sail's docs](https://laravel.com/docs/sail#configuring-a-shell-alias).

```shell
sail up
sail composer setup
sail npm run build 
```

This is a first pass of a very low level backend and frontend framework introduction without dropping any of the existing features:

**Backend**

- Laravel runs before all existing scripts and routes the requests accordingly - making sure headers are set via middlewares and the framework's authentication and session are bootstrapped (among other things like cookie encryption, CSP, ...)
- Adjusted Docker/Apache/nginx/Valet configurations to only use `public/index.php` as the application's entrypoint
- Introduce a separate "Legacy" domain within `app` for models for existing tables (controllers etc in the future)
- Move previous `index.php` to `home.php` to serve as the front page
- Wildcard `{file}.php` routes buffer script outputs (within `public` directory) and either pass it to a blade layout or respond with JSON (and appropriate header etc.)
- Replace cookie, connect and web api authentication with guards and user providers (cookie column can be removed in the future)
- Use configuration files (config() global helper) instead of accessing environment variables directly which will in turn allow to cache configuration in the future
- Replace global helpers with Laravel equivalents where applicable (request, asset, url, cookie) - still some cleanup to do
- Cronjobs are handle via Console Kernel (`app/Console/Kernel.php`) and scheduled accordingly
- Do not allow GET requests in public/request/**/*.php & POST request csrf protection
- Log Viewer for error monitoring (Sentry would be nice)
- Make phpunit tests, phpstan, and php cs fixer work
- No more single character input variables

**Database**

- Move existing sql migrations to legacy migrations (`database/migrations/legacy`) - migrations are safe to run on an existing database
- Add timestamps to `SetRequest` and `Subscription` tables (latter were already present on production, migration was missing)
- Drop unused tables: PlaylistVideo, ScoreHistory

**Frontend**

- Normalize UI elements
- Replace layout render functions with app layout and components; remove <head> and <body> tags from scripts
- Proper error pages via Laravel's exception handler: https://stage.retroachievements.org/nope
- Blade template components preparations https://stage.retroachievements.org/demo/
- Global XHR success and errors handlers as well as CSRF token header
- Enforce CSP rules
- Node pipeline using Vite
- Alpine.js and Tailwind (A and T of the TALL stack) added to the frontend stack - the existing js and css files are kept intact for now
- Theme and scheme selector using css variables (instead of loading css files)
- Minor increase of width by using tailwind's default lg breakpoint (1024px) - xl and 2xl are prepared
- Settings (controlpanel) as wide page example

**Content**

- Added Privacy Policy / Legal page: https://stage.retroachievements.org/terms
- Added cheevo to status/error pages:
	- https://stage.retroachievements.org/demo/error/400
	- https://stage.retroachievements.org/demo/error/401
	- https://stage.retroachievements.org/demo/error/403
	- https://stage.retroachievements.org/demo/error/404
	- https://stage.retroachievements.org/demo/error/405
	- https://stage.retroachievements.org/demo/error/406
	- https://stage.retroachievements.org/demo/error/408
	- https://stage.retroachievements.org/demo/error/409
	- https://stage.retroachievements.org/demo/error/413
	- https://stage.retroachievements.org/demo/error/414
	- https://stage.retroachievements.org/demo/error/415
	- https://stage.retroachievements.org/demo/error/419
	- https://stage.retroachievements.org/demo/error/429
	- https://stage.retroachievements.org/demo/error/500
	- https://stage.retroachievements.org/demo/error/501
	- https://stage.retroachievements.org/demo/error/502
	- https://stage.retroachievements.org/demo/error/503

**Deployment**

- Deploy script variant with maintenance mode (php artisan down)
- Change application version during deployment
- Updated deployment to allow rollbacks
- Build frontend during deployment

> **Note**
> This PR includes and installs dependencies that are not yet used but are prepared for future use (see composer.json)

- Fortify for authentication; providing a streamline authentication layer and two factor authentication
- Passport for OAuth server; providing personal access token management for the Web API and "Login with your RetroAchievements account" feature for third party services
- Horizon for queued jobs which will allow us to dispatch expensive tasks during requests
- Websockets for real-time notifications in the frontend
- Livewire for backend driven interactive components

> **Note**
> A lot of V2's usable prototype code has been parked outside of this PR or deactivated in code to not blow it out of proportion even more - will follow after this is merged

- Pretty URL routes have been commented out but can be reactivated one by one - a legacy URL 0redirector allows to redirect old `.php` to be redirected to the pretty URLs by merely removing the file from the `public` directory. The only way forward for v1 files is to be removed as functionality is replaced imo
- Readily made blade views for resource components (community, platform, site, ...) - not yet usable as everything is still handled via the legacy script layer. Components cannot be used within our scripts' `echo` outputs as that would cause an expensive intermediate step. Not worth it - let's replace those scripts down the line instead
- Actions that would modify V2 data (except for use in sync commands) are included but not yet used
- Overhauled frontend javascript which will replace jQuery and knockout
- Resource grid / table components to be reused for _any_ resource
- User settings for locale and language and applying translations

**Next up**

Some things I'll take care of next in separate PRs. I think this one's large enough /cough/.

- Migrate permissions to prepared role-based access control (RBAC)
- Finish V1 to V2 database direct and incremental sync support (`php artisan ra:sync`) via Laravel's scheduler (`app\Console\Kernel.php`) - it does work on my machine but i assume it still needs some tweaks to be run on production
- Populate user games relation
- [Laravel Octane](https://laravel.com/docs/sail) compatibility - will be used for connect and web api to increase performance
- Replace "legacy" mysqli database layer with Laravel's database connection manager (with appropriate sql mode settings and collation)
- Make V2's tests work again - the mysqli database connection within `lib/bootstrap.php` has to be completely removed first
- Remove superfluous DeletedModels table and delete v2 models in v1 directly to overcome sync issues
- Remove webserver-level rate limiting and configure Laravel's throttling capabilities instead
- Re-introduce Livewire and the prototype's abstract resource grid / table compoenents
- Re-introduce the prototype's integration and emulator release management features
- Replace public scripts with views and get rid of all those `echo` statements
- Split interceptor into middlewares and webhook notifications (discord)
- Move cron scripts to scheduler with commands in legacy domain
- Replace icon image assets with svg icons where applicable
- Replace header main menu with blade components
- Refactor remaining js assets into the Vite pipeline
- Use mail/templates & storage drivers
- Generate connect api docs
- ...
- ~Set up subdomains for static files, media files, Web API, Connect API~ Done.
- ~Replace layout header with blade components~ https://github.com/RetroAchievements/RAWeb/pull/1103/commits/f7fa22af702ae3a697e4d005ecc0c152f9d90592
- ~Refactor remaining css assets into the Vite pipeline~ https://github.com/RetroAchievements/RAWeb/pull/1103/commits/f407b85b10a1ce26e32449ed40d09e024a534bbe

**Caveats**

The learning curve for web development contributions gets steeper with this change due to added complexity. I cannot stress enough how much it'll be worth it though, from maintenance and security standpoints alone. Can't wait to drop a lot of our custom solutions and replace them with proper tools.

Bootstrapping the Laravel framework for each request increases response time. Our server is beefy enough to handle it though. Laravel Octane will eventually solve this by running select parts of the application in memory (some features have to be migrated to framework code first). However, database will always be the first bottleneck to run into.

node.js is actually required for development now, not only mentioned in the readme. It has to be run at least once when setting up the project and after each code update.

### CODE REVIEW

**Comment by luchaos** [comment]
Note: this PR will not be squashed; it will be closed and merged into the master branch to retain history.
Feel free to state PRs based on the v2 branch. The same rules apply as for the master branch.

File: public/request/ticket/create.php [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
+    'mode' => 'required|boolean',
+    'issue' => 'required|integer|min:1|max:2',
+    'note' => 'sometimes|string|max:2',
+]);
+
+$achievementID = (int) $input['achievement'];
+$problemType = $input['issue'];
+$hardcore = (int) $input['mode'];
+
+// TODO untangle $_POST['note']
```
Comment (Jamiras) [inline public/request/ticket/create.php:24]: is this related to the validation? it reads as if the note cannot be more than two characters long.

File: public/userInfo.php [hunk 2]
```diff
@@ -143,21 +128,14 @@
...
     "/user/$userPage",
-    "User page for $userPage"
-); ?>
-    <?php RenderTitleTag($pageTitle); ?>
-</head>
-<body>
-<?php RenderHeader($userDetails); ?>
+    "$userPage Profile"
+);
+?>
```
Comment (Jamiras) [inline public/userInfo.php:138]: doesn't this still need a `RenderContentStart`?

**Review by Jamiras — COMMENTED** [review]
Still have a few more things I want to look into once I can get the code running.

File: config/database.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            'driver' => 'mysql',
+            'url' => env('DATABASE_URL'),
+            'host' => env('APP_ENV') === 'local' && env('LARAVEL_SAIL') ? 'mysql' : env('DB_HOST', '127.0.0.1'),
+            'port' => env('APP_ENV') === 'local' && env('LARAVEL_SAIL') ? '3306' : env('DB_PORT', '3306'),
+            'database' => env('DB_DATABASE', 'forge'),
+            'username' => env('DB_USERNAME', 'forge'),
+            'password' => env('DB_PASSWORD', ''),
+            'unix_socket' => env('DB_SOCKET', ''),
+            'charset' => 'utf8mb4',
+            'collation' => 'utf8mb4_0900_ai_ci',
```
[the remark this answers is not in the document]
Comment (luchaos) [reply config/database.php:58]: Just noticed that mariadb does not support it - changed back to `utf8mb4_general_ci` as that's Laravel's default.

File: database/20220615_000000_Add_UserAccount_SoftcorePoints.sql [hunk 4]
```diff
@@ -1 +0,0 @@
-ALTER TABLE UserAccounts ADD COLUMN RASoftcorePoints INT DEFAULT 0 COMMENT 'Softcore Gamerscore' AFTER RAPoints;
```
[the remark this answers is not in the document]
Comment (luchaos) [reply database/20220615_000000_Add_UserAccount_SoftcorePoints.sql:1]: When starting with a fresh setup and migrating the database it would create the legacy tables from https://github.com/RetroAchievements/RAWeb/blob/v2/database/migrations/legacy/2012_10_03_133633_RA_Template.sql
I moved all the recent changes into that single sql file.
Starting with the previous development dump does not work anymore as it's wildly outdated. I'll generate a new dump as soon as the bigger setup issues have been resolved. The other migration sql scripts should not be necessary anymore.
That said, importing a prod db dump should not be necessary to run the application locally, the seeders should take care of that.
What I'd like to see as an alternative is that a local instance can pull data from prod via the web api instead to seed the local instance. Dumping and cleaning prod db seems a bit extreme to me - but it will be done soon. It's still on my list.

File: public/gamecompare.php [hunk 5]
```diff
@@ -91,6 +82,7 @@
             echo GetGameAndTooltipDiv($gameID, $gameTitle, $gameIcon, $consoleName, false, 96);
 
             echo "<form method=get action='/gamecompare.php'>";
+            echo csrf_field();
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/gamecompare.php:85]: Good catch, the csrf field, while not doing any harm in a GET from, is not required here. I went through all 74 occurrences of `<form>` to see if the csrf field should be applied to a POST form. I did not catch that one because it's missing the double quotes... I removed `method="get"` as that's the default method of forms.

File: public/leaderboardinfo.php [hunk 6]
```diff
@@ -50,28 +45,16 @@
...
     "/leaderboardinfo.php?i=$lbID",
     "Leaderboard: $lbTitle ($gameTitle, $consoleName): "
-); ?>
-    <?php RenderTitleTag($pageTitle); ?>
-</head>
-<body>
-<?php RenderHeader($userDetails); ?>
-
+);
+?>
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/leaderboardinfo.php:55]: Doesn't need it but it's much cleaner to have it in there 👍 
`RenderOpenGraphMetadata()` already takes care of setting the page title but `RenderContentStart` should still be present in case we'd want to wrap all the legacy scripts in some custom markup at some point.
I think the scripts will all be migrated to views and controllers before we'll ever need a wrapper. Clean is clean, added where missing.

File: public/request/comment/delete.php [hunk 7]
```diff
@@ -1,20 +1,19 @@
...
+if (!authenticateFromCookie($user, $permissions, $userDetails, Permissions::Registered)) {
+    abort(401);
+}
 
-$articleID = requestInput('a', 0, 'integer');
-$commentID = requestInput('c', 0, 'integer');
+$input = Validator::validate(request()->post(), [
+    'commentable_id' => 'required|integer',
+    'comment' => 'required|string|max:60000',
+]);
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/comment/delete.php:13]: Thanks! I think I've been working on that just before pushing the v2 branch - it's a copy/paste error.
Will be addressed with the next request stability improvement commit.

[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/comment/delete.php:12]: Refactored and simplified comment deletion logic.

File: public/leaderboardList.php [hunk 8]
```diff
@@ -377,36 +352,22 @@ function UpdateLeaderboard(user, lbID) {
...
-                echo "&#124;";
-                echo "&nbsp;";
-                echo $lbNumEntries . " entries";
-                echo "&nbsp;";
-                echo "&#124;";
-                echo "</div>";
-                if ($editAllowed) {
-                    echo "<div class='rightalign'><input type='submit' name='Update' onclick=\"UpdateLeaderboard('$user', '$lbID')\" value='Update'></div>";
-                }
+                echo " | ";
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/leaderboardList.php]: The JR Dev vs Dev paths should still be intact. I extracted the common parts and inverted the conditions but missed the remaining pipe. Cleaned up in another commit: https://github.com/RetroAchievements/RAWeb/pull/1103/commits/0b319c1d30df5e84d87eaa8ef4b0f1cdc7d424c1

File: public/gamecompare.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jamiras) [reply public/gamecompare.php]: >  while not doing any harm in a GET from
It breaks bookmarks and the ability to share a URL with another user. This is one page that we sometimes link for moderation (though usually just the landing page, not the reload after changing the selected game), so the ability to share the URL is fairly important.

File: public/request/ticket/create.php (hunk 1 above)
Comment (Jamiras) [reply public/request/ticket/create.php:24]: Verified via test.

>  The note must not be greater than 2 characters.

File: public/request/message/read.php [hunk 9]
```diff
@@ -1,19 +1,21 @@
...
-
-if (markMessageAsRead($user, $messageID, $messageReadStatus)) {
-    echo "OK:" . $messageID;
-} else {
-    echo "FAILED!";
+if (markMessageAsRead($user, $input['message'], $input['status'])) {
+    if ((int) $input['status'] === 1) {
+        return response()->json(['message' => __('legacy.success.ok')]);
+    }
+    return true;
```
[the remark this answers is not in the document]
Comment (Jamiras) [reply public/request/message/read.php:18]: Not sure if it's related (or even if it's new behavior), but opening an unread message does not remove the underline indicating that it's now been read. Refreshing the page does reflect the change.

File: resources/views/components/menu/footer.blade.php [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+            // [
+            //     'active' => Route::is('system*'),
+            //     'label' => __res('system'),
+            //     'url' => route('system.index'),
+            //     'visible' => Gate::allows('viewAny', App\Platform\Models\System::class),
+            // ],
+            [
+                'active' => Route::is('game*'),
+                'label' => __res('game'),
+                'url' => url('leaderboardList.php'),
```
Comment (Jamiras) [inline resources/views/components/menu/footer.blade.php:32]: this should be pointing at gameList.php

File: resources/views/components/menu/footer.blade.php [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+                'active' => Route::is('user*'),
+                'label' => __res('user'),
+                'url' => url('userList.php'),
+                // 'url' => route('user.index'),
+                'visible' => true,
+            ],
+            [
+                'active' => Route::is('news*'),
+                'label' => __res('news'),
+                'url' => url('developerstats.php'),
```
Comment (Jamiras) [inline resources/views/components/menu/footer.blade.php:97]: I'm not sure what this is supposed to be pointing at, it appears to be new.

Comment (luchaos) [reply resources/views/components/menu/footer.blade.php]: News index is prepared but does not exist yet - linking it developerstats was a mistake.

File: public/request/forum-topic-comment/update.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+if (!getSingleTopicComment($commentID, $commentData)) {
+    return back()->withErrors(__('legacy.error.error'));
+}
+
+if ($user != $commentData['Author'] && $permissions < Permissions::Admin) {
+    return back()->withErrors(__('legacy.error.permissions'));
+}
+
+if (editTopicComment($commentID, $commentPayload)) {
+    return back()->with('success', __('legacy.success.update'));
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/forum-topic-comment/update.php]: Changed to previous behaviour.

File: public/request/ticket/create.php (hunk 1 above)
Comment (luchaos) [reply public/request/ticket/create.php]: I was in the middle of rewriting the valdiation rules when I decided to push this branch anyways :) 
The validation rule made no sense yet nor did `note` being a nested form field, hence the TODO note.
Untangled here: https://github.com/RetroAchievements/RAWeb/commit/9acb69ad08d355edfc2f9fa2811dc5b8547b0602

File: public/request/message/read.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/message/read.php:18]: It's related - has to do with how the global XHR success and error handler work: https://github.com/RetroAchievements/RAWeb/blob/v2/public/js/all.js#L16-L38

To unify how "ajax" calls react to json responses I added those global handlers so info, success and error messages are displayed at any time that they are provided in a response.

The async "read" action would cause a success status message each time a message detail is opened. To prevent this I chose to return true which results in an empty yet successful HTTP response without triggering the status message. If that's desired it's easy to change. Error responses would still show every time.

The code that changed the looks of read messages on the fly was very convoluted. I did not want to spend too much time with refactoring it so I felt that seeing the correct status after a page reload is enough. I can see how that's a step back in visual feedback quality. I'll see if I can find a simple solution for that at another time; noted.

[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/message/read.php:15]: Cleaned up and visual feedback for read and unread is back: https://github.com/RetroAchievements/RAWeb/pull/1103/commits/17483067b981bf7f116a6e3b250534ef8277f2bc
Fixed another issue which prevented deletion of messages where the username case did not match (in my case "Luchaos" vs "luchaos")

File: public/request/user/update-relationship.php [hunk 13]
```diff
@@ -1,27 +1,19 @@
 <?php
 
-require_once __DIR__ . '/../../../vendor/autoload.php';
-require_once __DIR__ . '/../../../lib/bootstrap.php';
+if (!authenticateFromCookie($user, $permissions, $userDetail)) {
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/user/update-relationship.php:7]: Addressed here https://github.com/RetroAchievements/RAWeb/pull/1103/commits/bec126a394206c668f7c0fe7162f3f37e2e78d00

File: public/request/leaderboard/create.php [hunk 14]
```diff
@@ -1,35 +1,33 @@
...
-        $lbID = null;
-        if (submitNewLeaderboard($gameID, $lbID, $user)) {
-            // Good!
-            header("Location: " . getenv('APP_URL') . "/leaderboardList.php?g=$gameID&e=ok");
-            exit;
-        } else {
-            header("Location: " . getenv('APP_URL') . "/leaderboardList.php&e=cannotcreate");
-            exit;
-        }
+if (!authenticateFromCookie($user, $permissions, $userDetails, Permissions::JuniorDeveloper)) {
```
[the remark this answers is not in the document]
Comment (luchaos) [reply public/request/leaderboard/create.php:6]: Addressed in https://github.com/RetroAchievements/RAWeb/pull/1103/commits/89776a590a0f5411d9bf8ab3f8b1c55827b45c2a

File: database/20220615_000000_Add_UserAccount_SoftcorePoints.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (luchaos) [reply database/20220615_000000_Add_UserAccount_SoftcorePoints.sql:1]: > until then, shouldn't these stick around

you're right, they should. I went ahead and rewrote the sql files to a single sequential migration.
this should allow to have the tables all up to date no matter when the last sql file was executed.
it will not include cleanup sql statements however (points recalc, setting timestamps, etc) - those should not be done in schema migrations.
the file will follow in one of the next commits.

**Review by Jamiras — APPROVED** [review]

File: database/20220615_000000_Add_UserAccount_SoftcorePoints.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (luchaos) [reply database/20220615_000000_Add_UserAccount_SoftcorePoints.sql:1]: Done in https://github.com/RetroAchievements/RAWeb/pull/1103/commits/9c40f9cbdb3801188e20f438ce78e1bc35ba0ac3
